migrations/blocked_tasks.js

```diff
@@ -0,0 +1,74 @@
+function taskIsBlocked(task) {
+    for(dep of task.depends_on) {
+        if (dep.unattainable) {
+            return true
+        }
+    }
+
+    return false
+}
+
+function unattainableResolved(task) {
+    for (dep of task.depends_on) {
+        if (dep.unattainable == null) {
+            return false
+        }
+    }
+    return true
+}
+
+finishedStatuses = ["success","failed"]
+millisecondsPerDay = 86400000
+delta = millisecondsPerDay * 1
+
+startingDate = ISODate("2016-01-01T00:00:00Z")
+while(startingDate < new Date()) {
+    loops = 0
+    tasksSize = 0
+    while (true) {
+        tasks = db.tasks.find({"create_time": {"$gte": startingDate, "$lte": new Date(startingDate.getTime() + delta)}, "status":"undispatched", "depends_on.status": {"$in": ["success", "failed", "", "*"]}, "depends_on": {"$elemMatch":{"unattainable": {"$exists": false}}}}, {"depends_on":1}).toArray()
+        printjson(loops++)
+        if (tasks.length == 0 || tasks.length == tasksSize) {
+            break
+        }
+        tasksSize = tasks.length
+
+        for (i=0; i < tasks.length; i++) {
+            taskUpdated = false
+            dependsOn = tasks[i].depends_on
+            for (j=0; j < dependsOn.length; j++) {
+                if(dependsOn[j].unattainable != null) {
+                    continue
+                }
+                
+                depTask = db.tasks.findOne({"_id": dependsOn[j]._id}, {"status": 1, "depends_on": 1});
+                if(depTask == null) {
+                    printjson(tasks[i])
+                    break
+                }
+
+                if (finishedStatuses.includes(depTask.status)) {
+                    taskUpdated = true
+                    // 1st degree blocked
+                    if(depTask.status != "*" && depTask.status != dependsOn[j].status) {
+                        dependsOn[j].unattainable = true
+                    } else {
+                        dependsOn[j].unattainable = false
+                    }
+                } else if (dependsOn[j].status != "*" && taskIsBlocked(depTask)) {
+                    taskUpdated = true
+                    dependsOn[j].unattainable = true
+                } else if (unattainableResolved(depTask)) {
+                    taskUpdated = true
+                    dependsOn[j].unattainable = false
```

Comment:
`unattainableResolved` means that every dependency is unattainable. Does this mean we should set this to `true`? I might be misreading this though.

Reply:
By `unattainableResolved` I meant we've resolved for each dependency of a task if it is unattainable or not. 
So if _A_ depends on _B_ which depends on _C_ and _C_ fails, if we look at _A_ before we look at _B_ we can't yet know for certain if _B_ didn't run because it is blocked (blocking _A_), so we will leave _A_ for this iteration of the inner while loop and come back to it after we've updated _B_. If they are resolved, however, then we know _B_ wasn't blocked and didn't run for some other reason and we can safely say that _A_'s dependency on _B_ is not unattainable.